src/main/java/org/javarosa/core/model/TriggerableDag.java

```diff
@@ -516,9 +523,15 @@ private Set<QuickTriggerable> getAllToTrigger(Set<QuickTriggerable> cascadeRoots
     private Set<QuickTriggerable> doEvaluateTriggerables(FormInstance mainInstance, EvaluationContext evalContext, Set<QuickTriggerable> toTrigger,
                                                          TreeReference changedRef, Set<QuickTriggerable> affectAllRepeatInstances, Set<QuickTriggerable> alreadyEvaluated) {
         Set<QuickTriggerable> evaluated = new HashSet<>();
+
         EvaluationContext context;
         if (predicateCaching) {
-            context = new EvaluationContext(evalContext, new IdempotentInMemPredicateCache());
+            List<PredicateFilter> filters = Stream.concat(
+                customPredicateFilters.stream(),
+                Stream.of(indexPredicateFilter, cachingPredicateFilter, new IdempotentPredicateCache())
```

Comment:
> Not sure there's action to take here but did want to flag it. Maybe the last cache type's header comment should include that it's only intended for use in a single evaluation cascade. Another idea would be to build a single cache and explicitly reset it at the beginning or end of each cascade evaluation.

There is currently a comment on `IdempotentPredicateCache`:

> Caches down stream evaluations (in the {@link PredicateFilter} chain) for "idempotent" (with respect to current form state) predicates. Can only be used for static instances or in cases where form state won't change - will cause clashes otherwise. Repeated evaluations are fetched in O(1) time.

Does that feel like enough or do you think it needs more details?

Reply:
> Side note, why does similarCmpAndEqExpressionsDoNotGetConfused fail in that case? Are there cases where the expressions could get confused within a single cascade?

Not sure! The cache definitely isn't designed to work between evaluation runs, so I'd imagine that's messing with something.